### PR TITLE
large_assignments: Lint on specific large args passed to functions

### PR DESCRIPTION
**Pull request by Enselic**
Requires lowering function call arg spans down to MIR, which is done in the second commit.

Part of #83518

Also see 
* https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/arg.20Spans.20for.20TerminatorKind.3A.3ACall.3F
* https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/move_size_limit.20lint

r? @oli-obk (E-mentor)

### CODE REVIEW

File: compiler/rustc_middle/src/thir.rs
Comment (oli-obk) [inline compiler/rustc_middle/src/thir.rs:292]: Can't you get the spans from the expr that the ExprId in the args above refers to?

And from that, you can eliminate all the thir and hir changes done in this PR?

File: compiler/rustc_middle/src/mir/syntax.rs
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/syntax.rs:673]: Could also turn this into a `Vec<Spanned<Operand<'tcx>>>`

**Review by oli-obk — CHANGES_REQUESTED** [review]
I only reviewed the first commit. I don't understand why you needed to change HIR types to be !Copy. Is that a leftover from previous experimentation adding such a Vec to the HIR?

File: compiler/rustc_middle/src/thir.rs
Comment (Enselic) [reply compiler/rustc_middle/src/thir.rs]: Indeed I can! When I tried to do it before, I couldn't figure out how to do it. But now when I tried again, it was easy, and makes the code much simpler. You were also right about !Copy being a left-over from adding arg_spans to HIR early on, and I forgot about it. Removed again now.

File: compiler/rustc_middle/src/mir/syntax.rs
Comment (Enselic) [reply compiler/rustc_middle/src/mir/syntax.rs:673]: I tried, but to do that we need to make Spanned derive TypeFoldable and TypeVisitable (which TerminatorKind already does), and to do that we need to make rustc_span depend on rustc_middle, and I don't think we want that?

File: compiler/rustc_middle/src/mir/syntax.rs
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/syntax.rs:675]:
```suggestion
        /// (e.g. `a` and `b` in `x.foo(a, b)`).
        /// Has the same length as `args`.
```

File: compiler/rustc_middle/src/mir/syntax.rs
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/syntax.rs:673]: You should be able to add a manual impl to rustc_middle. We already have a bunch of them there. Or ad a generic impl to rustc_span in case that already depends on rustc_type_ir

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/syntax.rs:673]: Worst case make it a `Vec<(Span, Operand)>`

File: compiler/rustc_middle/src/mir/syntax.rs
Comment (Enselic) [reply compiler/rustc_middle/src/mir/syntax.rs:675]: I agree this would have been good to add, but now that I started using `Spanned<Operand<'tcx>>` there is nothing that can get out of sync any longer.

File: compiler/rustc_middle/src/mir/syntax.rs
Comment (Enselic) [reply compiler/rustc_middle/src/mir/syntax.rs]: Oh. Right. Done! (`Spanned`.)

**Comment by Enselic** [comment]
Looks like I need to bless 309 mir-opt tests to make CI pass. To not make the diff too noisy, would you mind doing a code review of the first commit before I bless tests? Am I on the right track?

@rustbot ready

**Review by oli-obk — COMMENTED** [review]
I did write a review last night but seem to have fallen asleep while doing so.

None of this is important or even nits you must address, pick what you like

I like how it works now

**Comment by oli-obk** [comment]
Ah, I think we shouldn't bless the mir opt tests. Instead change the mir printing logic to skip the Spanned and directly print the Operand

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
Comment (Enselic) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:3182]: Agreed, I think this is a left-over from when I was still figuring out the best way to do things. Will fix.

File: compiler/rustc_codegen_ssa/src/mir/block.rs
Comment (Enselic) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:867]: Agreed, I think this is a left-over from when I was still figuring out the best way to do things. Will fix.

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs
Comment (Enselic) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs:125]: I don't think I can inside the `map()`? I get
> the `?` operator can only be used in a closure that returns `Result` or `Option`

File: compiler/rustc_borrowck/src/lib.rs
Comment (Enselic) [reply compiler/rustc_borrowck/src/lib.rs:718]: I actually tried that at first, but discovered that sometimes `consume_operand()` is called in contexts where we don't have a `Spanned<Operand>`, only an `Operand`, and I didn't want to have to introduce `DUMMY_SP`. But this is definitely something that might make sense to do in some follow-up PR that does further refactoring.

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs
Comment (oli-obk) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs:125]: This closure does return a Result

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
Comment (Enselic) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs]: Done

File: compiler/rustc_codegen_ssa/src/mir/block.rs
Comment (Enselic) [reply compiler/rustc_codegen_ssa/src/mir/block.rs]: Done

**Comment by Enselic** [comment]
@RalfJung Do you think this (`Spanned<Operand>>`) looks good? If both you and @oli-obk likes it, wouldn't it be nice to later make a PR that changes `TerminatorKind::Drop` from

    func: Operand<'tcx>,
    fn_span: Span,
    ...

to

    func: Spanned<Operand<'tcx>>,
    ...

? I would be happy to do that later if you want it done.

**Comment by Enselic** [comment]
> Instead change the mir printing logic to skip the Spanned and directly print the Operand

Done!

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs
Comment (Enselic) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: Indeed. Done!

**Comment by RalfJung** [comment]
The change in compiler/rustc_middle/src/mir/syntax.rs LGTM, assuming this doesn't cost too much perf. We should do a perf run before landing. Let me know when you're ready for that, I can tell the bots to do it.

> wouldn't it be nice to later make a PR that changes TerminatorKind::Drop

No strong opinion, but sure, why not. :)
We can even almost make `ConstOperand` just a `Spanned<Const>`, but sadly there's this user_ty field getting in the way...

File: compiler/rustc_codegen_cranelift/src/abi/mod.rs
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/abi/mod.rs:363]: Maybe use `operand.node` on the call sites of `codegen_call_argument_operand()` instead?

File: compiler/rustc_monomorphize/src/collector.rs
Comment (bjorn3) [inline compiler/rustc_monomorphize/src/collector.rs:677]: This doesn't check if the operand is a move operand.

File: compiler/rustc_codegen_cranelift/src/abi/mod.rs
Comment (Enselic) [reply compiler/rustc_codegen_cranelift/src/abi/mod.rs]: Sure, why not. Done!

File: compiler/rustc_monomorphize/src/collector.rs
Comment (Enselic) [reply compiler/rustc_monomorphize/src/collector.rs:677]: Yep, this is a known task which I added to the [tracking issue ](https://github.com/rust-lang/rust/issues/83518) a few days ago after RalfJung pointed out the same thing. I plan on fixing it relatively soon in a future PR.

**Comment by Enselic** [comment]
So, turns out this impacts performance. At first I was sad. Then I started to think that maybe this is inevitable. We do more work when lowering to MIR now. So it would be strange if that didn't impact performance. I guess the question is what magnitude of slow-down is acceptable. A mean of 0.5% does not seem that bad to me. But I understand if that makes everyone hesitate.

Not sure yet how to improve performance of this PR. I think the first step is to see if I can reproduce the perf regression locally with https://github.com/rust-lang/rustc-perf and then try to find the exact bottleneck via detailed benchmarking.

Before I spend time on doing that though I would like to ask what your thoughts are on the performance regression numbers?

**Comment by RalfJung** [comment]
A .5% mean regression across 47 primary benchmarks is non-trivial, that's a lot worse than what we see most weeks I think.

Why does this cost any performance when the lint is not enabled? That really shouldn't be the case.

**Comment by oli-obk** [comment]
> Why does this cost any performance when the lint is not enabled? That really shouldn't be the case.

The cost is entirely in tracking the new `Span`s in MIR I think:

```
  8,977,505  ???:<rustc_span::span_encoding::Span as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
  8,926,092  ???:<rustc_span::span_encoding::Span as rustc_serialize::serialize::Encodable<rustc_middle::query::on_disk_cache::CacheEncoder>>::encode
  7,159,141  ???:<rustc_middle::mir::BasicBlockData as rustc_data_structures::stable_hasher::HashStable<rustc_query_system::ich::hcx::StableHashingContext>>::hash_stable
 -5,365,960  ???:<alloc::vec::Vec<rustc_middle::mir::syntax::Operand> as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
  5,308,607  ???:<hashbrown::raw::RawTable<(rustc_span::span_encoding::Span, usize)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::span_encoding::Span, usize, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
  4,425,396  ???:<rustc_span::span_encoding::Span as rustc_serialize::serialize::Encodable<rustc_metadata::rmeta::encoder::EncodeContext>>::encode
  2,988,810  ???:<rustc_middle::mir::BasicBlockData as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
  2,881,802  ???:<rustc_data_structures::sip128::SipHasher128>::short_write_process_buffer::<8>
  2,547,835  ???:<rustc_middle::mir::syntax::Operand as rustc_serialize::serialize::Encodable<rustc_middle::query::on_disk_cache::CacheEncoder>>::encode
  2,327,418  ???:<rustc_middle::mir::syntax::Operand as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
  2,155,701  ???:<rustc_span::span_encoding::Span as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
  1,976,462  ???:<rustc_span::caching_source_map_view::CachingSourceMapView>::span_data_to_lines_and_cols
  1,905,733  ???:rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>
 -1,905,733  ???:rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_hir::hir_id::OwnerId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>
 -1,750,003  ???:<rustc_middle::mir::Body as rustc_serialize::serialize::Encodable<rustc_middle::query::on_disk_cache::CacheEncoder>>::encode
 -1,571,356  ???:<hashbrown::raw::RawTable<(rustc_middle::ty::context::InternedInSet<rustc_middle::ty::list::List<rustc_middle::ty::generic_args::GenericArg>>, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_middle::ty::context::InternedInSet<rustc_middle::ty::list::List<rustc_middle::ty::generic_args::GenericArg>>, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
  1,571,334  ???:<hashbrown::raw::RawTable<(rustc_middle::ty::context::InternedInSet<rustc_middle::ty::list::List<rustc_middle::ty::Ty>>, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_middle::ty::context::InternedInSet<rustc_middle::ty::list::List<rustc_middle::ty::Ty>>, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
  1,483,859  ???:rustc_span::hygiene::decode_syntax_context::<rustc_middle::query::on_disk_cache::CacheDecoder, <rustc_span::hygiene::SyntaxContext as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode::{closure
  1,331,356  ???:<hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
 -1,331,323  ???:<hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, alloc::vec::Vec<rustc_hir::hir::TraitCandidate>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, alloc::vec::Vec<rustc_hir::hir::TraitCandidate>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
 -1,158,384  ???:<rustc_data_structures::sip128::SipHasher128>::finish128
  1,100,220  ???:<rustc_query_impl::query_impl::codegen_fn_attrs::dynamic_query::{closure
 -1,043,652  ???:<hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
  1,014,956  ???:<hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
    963,379  ???:<rustc_monomorphize::collector::MirUsedCollector as rustc_middle::mir::visit::Visitor>::visit_operand
   -947,335  ???:<rustc_span::span_encoding::Span>::to
    900,572  ???:<rustc_parse::parser::Parser>::parse_stmt_without_recovery
    798,336  ???:<rustc_data_structures::sip128::SipHasher128>::short_write_process_buffer::<4>
   -781,069  /build/glibc-sMfBJT/glibc-2.31/string/../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:__memcpy_avx_unaligned_erms
    709,190  ???:_rjem_je_arena_cache_bin_fill_small
   -651,592  ???:<rustc_parse::parser::Parser>::parse_outer_attributes
    503,040  ???:<hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
   -484,290  ???:<hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_resolve::Module)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_resolve::Module, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
    484,280  ???:<hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
   -469,894  ???:<hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
    445,655  ???:rustc_interface::callbacks::track_span_parent
    439,765  ???:<rustc_span::hygiene::SyntaxContext as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
    388,145  ???:<rustc_data_structures::sip128::SipHasher128>::short_write_process_buffer::<1>
```

**Comment by RalfJung** [comment]
Ah... Yeah that does seem hard to avoid.

And for the initially built MIR, this is redundant, right? All arguments are temporaries. We could look for their last assignment in the current bb and use the span of the RHS?

**Comment by oli-obk** [comment]
Almost. We discussed this in https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/arg.20Spans.20for.20TerminatorKind.3A.3ACall.3F/near/393101506 and @Enselic showed

```rust
fn main() {
    let data = [0; 9999]; // line 2
    analyze(data);        // line 3
}

pub fn analyze(_data: [u8; 9999]) {}
```

as an example where we can't possibly get the span of the `data` in argument position (I think we are not generating a temporary there, but maybe we are?).

**Comment by Enselic** [comment]
For that example code it looks like temporaries are used with  `-Zmir-opt-level=0`  but not with  `-Zmir-opt-level=1` (see below). Maybe we should move analysis so it always is performed on the unoptimized MIR? Then we might not have to lower `Spanned` args down to MIR, because we would have temporaries from which we can figure out `Span`. We want to move analysis out of rustc_monomorphize collector.rs anyway.  And having users relying on the optimizer to "fix" large moves seems risky and worth linting for.

```console
$ rustc +nightly --emit=mir -o - -Zmir-opt-level=0 main.rs
fn main() -> () {
    let mut _0: ();
    let _1: [u8; 9999];
    let _2: ();
    let mut _3: [u8; 9999];
    scope 1 {
        debug data => _1;
    }

    bb0: {
        StorageLive(_1);
        _1 = [const 0_u8; 9999];
        StorageLive(_2);
        StorageLive(_3);
        _3 = _1;
        _2 = analyze(move _3) -> [return: bb1, unwind continue];
    }
}
$ rustc +nightly --emit=mir -o - -Zmir-opt-level=1 main.rs
fn main() -> () {
    let mut _0: ();
    let _1: [u8; 9999];
    let _2: ();
    scope 1 {
        debug data => _1;
    }

    bb0: {
        _1 = [const 0_u8; 9999];
        _2 = analyze(_1) -> [return: bb1, unwind continue];
    }
}
```

**Comment by oli-obk** [comment]
you need to dump the MIR after building to see the right MIR.

But the MIR you see during the lint won't contain this anymore since it happens after optimizations.

So... one more thing we could try is to move back to your original design of having a separate list of spans for terminator arguments, but this time put it into a `ClearCrossCrate`, so it will not get serialized to metadata. From the cachegrind dump it looks like this is part of the regression.

**Comment by oli-obk** [comment]
> I don't understand, but this seems important to understand. Are you saying `--emit=mir ` gives inaccurate MIR output when combined with `-Zmir-opt-level=...` like in my example commands?

It gives you the MIR after optimizations, which is what you are seeing  for the analysis. The mir after mir building is mostly the same no matter the opt level.



> Not if we move the analysis to before optimizations as I proposed, right?

that is not really possible. We need to do it after monomorphization to get actual memory usage of actual places that copy large things.

**Comment by oli-obk** [comment]
https://github.com/rust-lang/rust/pull/116927 did not show a relevant performance change. So the entire regression is in incremental cache writing, which we cannot avoid.

I don't think we can do anything here other than regress perf, or use the suboptimal diagnostics and live with them, but maybe someone from @rust-lang/wg-compiler-performance  has ideas?

**Comment by nagisa** [comment]
With the regression being at just 0.5%, is it really something to worry about that much? If we want to have this lint, it sounds to me like having reasonable diagnostics is perhaps non-negotiable requirement, but I also haven't seen a comparison between diagnostics that would be plausible without the argument spans.

**Comment by Kobzol** [comment]
The regression is not trivial, but I think that we could bite the bullet here if we deem the lint to be  worth it (which I think it might be).

I wonder where does the [metadata size increase](https://perf.rust-lang.org/compare.html?start=96c4dba79407dd92eaef8e8442fb4bec6d734bbc&end=11f7c417b8f495320ddf4cff8a4f7dbbe84c6184&stat=size%3Acrate_metadata) come from though? It's present also for non-incremental benchmarks.

**Comment by Enselic** [comment]
> I wonder where does the [metadata size increase](https://perf.rust-lang.org/compare.html?start=96c4dba79407dd92eaef8e8442fb4bec6d734bbc&end=11f7c417b8f495320ddf4cff8a4f7dbbe84c6184&stat=size%3Acrate_metadata) come from though?

Isn't it simply because we add fn arg spans to metadata? In https://github.com/rust-lang/rust/pull/116927 I used ClearCrossCrate which IIRC avoids increasing metadata size.  Maybe it doesn't matter but that PR had slightly better perf (0.4% regression vs the 0.5% regression on this PR).

**Comment by RalfJung** [comment]
We could test that hypothesis by moving the span tracking for arguments into a separate PR that we land first.

**Comment by Enselic** [comment]
Thanks for taking a look! Here is an isolated PR of just the lowering of fn arg spans down to MIR in a separate and rebased PR as requested [here](https://github.com/rust-lang/rust/pull/116520#issuecomment-1887418073) and [here](https://github.com/rust-lang/rust/pull/116520#issuecomment-1887390394): https://github.com/rust-lang/rust/pull/119880

**Comment by Enselic** [comment]
Benchmark numbers are in, and lowering fn arg spans to MIR regresses performance by a mean of [0.4%](https://github.com/rust-lang/rust/pull/119880#issuecomment-1889722218).

The consensus seems to be that this regression is worth it, so would be great if someone could r+ before new conflicts arise.

> I also haven't seen a comparison between diagnostics that would be plausible without the argument spans.

Maybe I misunderstand you; but without doing this lowering, we have to lint large moves into functions on the whole function call, so there is no sensible way to point out precisely which arg is too big. See [this](https://github.com/rust-lang/rust/pull/116520/files#diff-9a6d93699fdaf79fa88e80fc200089df7ac0632ab31c41fafc256dbf99e3c67a) diff for an example of how doing this lowering allows us to make diagnostics more accurate.

**Comment by oli-obk** [comment]
yea, let's land this. r=me after a rebase

@bors delegate+ rollup=never

**Comment by Enselic** [comment]
👍 @bors r=@oli-obk

**Comment by oli-obk** [comment]
This regression is expected.

We are tracking more spans in order to be able to produce correct diagnostics. We expect to be using this information more now that we have it. I do expect some improvements to be possible in borrowck errors and const prop lints

@rustbot label: +perf-regression-triaged

**Pull request merged**
